packages/turbo/bin/turbo

```diff
@@ -1,10 +1,257 @@
+    console.error();
+    console.error('***');
+    console.error();
+    console.error(`We were not able to find the binary at:\n${correctBinary}`);
+    console.error();
+    console.error(`We looked for it at:`);
+    console.error(require.resolve.paths(correctBinary).join('\n'));
+  }
+  if (alternateBinary !== null) {
+    console.error();
+    console.error('***');
+    console.error();
+    console.error(`Your platform (${platform}) can sometimes run x86 under emulation.`);
+    console.error(`We did not find a possibly-compatible binary at:\n${alternateBinary}`);
+    console.error();
+    console.error(`We looked for it at:`);
+    console.error(require.resolve.paths(alternateBinary).join('\n'));
+  }
+
+  // Investigate other failure modes.
+
+  // Has the wrong platform's binaries available.
+  const availableBinaries = availablePlatforms.map(platform => availableArchs.map(arch => `turbo-${platform}-${arch}/bin/turbo${platform === 'windows' ? '.exe': ''}`)).flat();
+  const definitelyWrongBinaries = availableBinaries.filter(binary => binary !== correctBinary || binary !== correctBinary);;
+  const otherInstalled = definitelyWrongBinaries.filter(binaryPath => {
+    try {
+      return require.resolve(binaryPath);
+    } catch (e) {}
+  });
+
+  console.error();
+  console.error('***');
+  console.error();
+
+  if (otherInstalled.length > 0) {
+    console.error('Turborepo checked to see if binaries for another platform are installed.');
+    console.error('This typically indicates an error in sharing of pre-resolved node_modules across platforms.');
+    console.error('One common reason for this is copying files to Docker.');
+    console.error();
+    console.error(`We found these unnecessary binaries:`);
+    console.error(otherInstalled.join('\n'));
+  } else {
+    console.error(`We did not find any binaries on this system.`);
+    console.error(`This can happen if you run installation with the --no-optional flag.`);
+  }
+
+  // Check to see if we have partially-populated dependencies in the npm lockfile.
+  const MAX_LOOKUPS = 10;
+  const availablePackages = availablePlatforms.map(platform => availableArchs.map(arch => `turbo-${platform}-${arch}`)).flat();
+
+  try {
+    // Attempt to find project root.
+    const selfPath = require.resolve('turbo/package');
+
+    let previous = null;
+    let current = path.join(selfPath, '..', 'package-lock.json');
+
+    for (let i = 0; previous !== current && i < MAX_LOOKUPS; i++) {
+      try {
+        const lockfile = fs.readFileSync(current);
+        const parsedLockfile = JSON.parse(lockfile);
+
+        // If we don't show up in the lockfile it's the wrong lockfile.
+        if (parsedLockfile?.dependencies?.turbo) {
+          // Check to see if all of `turbo-<PLATFORM>-<ARCH>` is included.
+          const hasAllPackages = Object.keys(parsedLockfile?.dependencies ?? {}).filter(dependency => availablePackages.includes(dependency)).length === availablePackages.length;
```

Comment:
@chris-olszewski to double-check me here.

I'm trying to _approximate_ "this is depended upon by the project." Are these JSON paths going to do what I want them to do?

Reply:
The JSON paths you're using will work for npm 5-8.

The `dependencies` field is a [legacy field](https://docs.npmjs.com/cli/v9/configuring-npm/package-lock-json#dependencies) that is used to allow older version of npm to read newer lockfiles. As of 9.0.0 `npm` will no longer populate this field by default. If we want to support lockfile versions 1-3 then we'd need to read `packages` and fallback to `dependencies` if it isn't present.

The keys of `packages` are slightly different which makes our lives harder. As long as the user is only specifying one version of `turbo`across the monorepo then the keys will just be `node_modules/turbo` etc. If they have multiple conflicting versions of turbo then they can be nested in workspace paths e.g. `apps/web/node_modules/turbo` and `node_modules/turbo`. 

Checking for the existence of top level `node_modules/turbo*` packages should be sufficient to since if they created a bad lockfile any missing nested turbo packages would also show up in the top level.

Reply:
I've updated the check to do the fallback (it actually falls back to do the 9+ check). It even cleaned up the code nicely!